### PR TITLE
docs: add key manager docs

### PR DESCRIPTION
**Pull request by SWvheerden**
Adds/updates and fixes documentation for the key manager service

### CODE REVIEW

File: base_layer/wallet/src/key_manager_service/handle.rs [hunk 1]
```diff
@@ -35,7 +35,8 @@ use crate::key_manager_service::{
     KeyManagerInner,
     KeyManagerInterface,
 };
-
+/// Struct used to create an easily and cheaply cloneable handle to the Key manager services that can be safely shared
+/// across multiple threads.
```
Comment (sdbondi) [inline base_layer/wallet/src/key_manager_service/handle.rs]: Perhaps a bit of detail on what the key manager does and how it works. 
```suggestion
/// The key manager provides a hierarchical key derivation function (KDF) that derives uniformly random secret keys from a single seed key for arbitrary branches, using an implementation of `KeyManagerBackend` to store the current index for each branch.
/// 
/// This handle can be cloned cheaply and safely shared across multiple threads.
```

File: base_layer/wallet/src/key_manager_service/handle.rs [hunk 2]
```diff
@@ -44,6 +45,9 @@ pub struct KeyManagerHandle<TBackend> {
 impl<TBackend> KeyManagerHandle<TBackend>
 where TBackend: KeyManagerBackend + 'static
 {
+    /// Creates a new Key manager and wrap this inside of the handle
+    /// * `master_seed` is the primary seed that will be used to derive all unique branch keys with their indexs
+    /// * `db` is the key manager backed to be used for persistent storage of branches and indices.
```
Comment (sdbondi) [inline base_layer/wallet/src/key_manager_service/handle.rs]: The fact that it wraps something else is an implementation detail so I would leave it out. 
A handle in Tari usually refers to something that sends messages to a service, so it should probably be renamed at some point in another PR. 
```suggestion
    /// Creates a new key manager.
    /// * `master_seed` is the primary seed that will be used to derive all unique branch keys with their indexes
    /// * `db` implements `KeyManagerBackend` and is used for persistent storage of branches and indices.
```

File: base_layer/wallet/src/key_manager_service/initializer.rs [hunk 3]
```diff
@@ -36,6 +36,7 @@ use crate::key_manager_service::{
     KeyManagerHandle,
 };
 
+/// Struct used to initialize the Key manager service using the [ServiceInitializer]
```
Comment (sdbondi) [inline base_layer/wallet/src/key_manager_service/initializer.rs]: nit: It's redundant to say that something is a struct in docs.
```suggestion
/// Initializes the key manager service by implementing the [ServiceInitializer] trait.
```

File: base_layer/wallet/src/key_manager_service/interface.rs [hunk 4]
```diff
@@ -25,6 +25,7 @@ use tari_common_types::types::PrivateKey;
 
 use crate::key_manager_service::error::KeyManagerServiceError;
 
+/// Wrapper struct to return results of adding a new branch to track
```
Comment (sdbondi) [inline base_layer/wallet/src/key_manager_service/interface.rs]:
```suggestion
/// The value returned from [add_new_branch]. `AlreadyExists` is returned if the branch was previously created, otherwise `NewEntry` is returned.
```

File: base_layer/wallet/src/key_manager_service/interface.rs [hunk 5]
```diff
@@ -36,28 +37,40 @@ pub struct NextKeyResult {
...
 #[async_trait::async_trait]
 pub trait KeyManagerInterface: Clone + Send + Sync + 'static {
+    /// Creates a new branch for the key manager service to track
+    /// If this is an existing branch, that is not yet tracked in memory, the key manager service will load the key
+    /// manager from the backend to track in memory, will return `Ok(AddResult::NewEntry)`. If the branch is already
+    /// tracked in memory the result will be `Ok(AddResult::AlreadyExists)`. If the branch does not exist in memory
+    /// or in the backend, a new branch will be created and tracked the backend, `Ok(AddResult::NewEntry)`.
     async fn add_new_branch<T: Into<String> + Send>(&self, branch: T) -> Result<AddResult, KeyManagerServiceError>;
 
+    /// Encrypts the sensitive data in the backend using the provided cipher
```
Comment (sdbondi) [inline base_layer/wallet/src/key_manager_service/interface.rs]:
```suggestion
    /// Encrypts the key manager state using the provided cipher. An error is returned if the state is already encrypted.
```

File: base_layer/wallet/src/key_manager_service/interface.rs [hunk 6]
```diff
@@ -36,28 +37,40 @@ pub struct NextKeyResult {
...
     async fn apply_encryption(&self, cipher: Aes256Gcm) -> Result<(), KeyManagerServiceError>;
 
+    /// Decrypts the backend data.
```
Comment (sdbondi) [inline base_layer/wallet/src/key_manager_service/interface.rs]:
```suggestion
    /// Decrypts the key manager state using the provided cipher. An error is returned if the state is not encrypted.
```

File: base_layer/wallet/src/key_manager_service/interface.rs [hunk 7]
```diff
@@ -36,28 +37,40 @@ pub struct NextKeyResult {
...
     async fn remove_encryption(&self) -> Result<(), KeyManagerServiceError>;
 
+    /// Gets the next key from the branch. This will auto increment the key index by 1
```
Comment (sdbondi) [inline base_layer/wallet/src/key_manager_service/interface.rs]:
```suggestion
    /// Gets the next key from the branch. This will auto-increment the branch key index by 1
```

File: base_layer/wallet/src/key_manager_service/storage/database/backend.rs [hunk 8]
```diff
@@ -24,14 +24,15 @@ use aes_gcm::Aes256Gcm;
 use crate::key_manager_service::{error::KeyManagerStorageError, storage::database::KeyManagerState};
 
 /// This trait defines the required behaviour that a storage backend must provide for the Key Manager service.
-/// Data is passed to and from the backend via the [DbKey], [DbValue], and [DbValueKey] enums. If new data types are
-/// required to be supported by the backends then these enums can be updated to reflect this requirement and the trait
-/// will remain the same
 pub trait KeyManagerBackend: Send + Sync + Clone {
+    /// This will retrieve the key manager spesified by the branch string, None is returned if no key manager is found
```
Comment (sdbondi) [inline base_layer/wallet/src/key_manager_service/storage/database/backend.rs]:
```suggestion
    /// This will retrieve the key manager specified by the branch string, None is returned if the key manager is not found
```

File: base_layer/wallet/src/key_manager_service/storage/database/backend.rs [hunk 9]
```diff
@@ -24,14 +24,15 @@ use aes_gcm::Aes256Gcm;
...
+    /// for the branch.
     fn get_key_manager(&self, branch: String) -> Result<Option<KeyManagerState>, KeyManagerStorageError>;
+    /// This will add an additional branch for the key manager to track.
     fn add_key_manager(&self, key_manager: KeyManagerState) -> Result<(), KeyManagerStorageError>;
+    /// This will increase the key index of the spesified branch, will return an error if the branch does not exist.
```
Comment (sdbondi) [inline base_layer/wallet/src/key_manager_service/storage/database/backend.rs]:
```suggestion
    /// This will increase the key index of the specified branch, and returns an error if the branch does not exist.
```

File: base_layer/wallet/src/key_manager_service/storage/database/mod.rs [hunk 10]
```diff
@@ -93,6 +103,7 @@ where T: KeyManagerBackend + 'static
             .and_then(|inner_result| inner_result)
     }
 
+    /// Decrypts the entite key manager.
```
Comment (sdbondi) [inline base_layer/wallet/src/key_manager_service/storage/database/mod.rs]:
```suggestion
    /// Decrypts the entire key manager.
```

File: base_layer/wallet/src/key_manager_service/storage/sqlite_db/key_manager_state.rs [hunk 11]
```diff
@@ -35,6 +35,7 @@ use crate::{
     },
 };
 
+/// Struct used to map directly to database
```
Comment (sdbondi) [inline base_layer/wallet/src/key_manager_service/storage/sqlite_db/key_manager_state.rs]:
```suggestion
/// Represents a row in the key_manager_states table.
```

File: base_layer/wallet/src/key_manager_service/storage/sqlite_db/key_manager_state.rs [hunk 12]
```diff
@@ -85,17 +88,22 @@ impl NewKeyManagerStateSql {
 }
 
 impl KeyManagerStateSql {
+    /// Retrieve every key manager branch currently in the database.
+    /// Will always return a vec of [KeyManagerStateSql], if none are found, it will return an empty vec
```
Comment (sdbondi) [inline base_layer/wallet/src/key_manager_service/storage/sqlite_db/key_manager_state.rs]: I don't think this is quite correct, it will return an error if the load fails.
```suggestion
    /// Returns a `Vec` of [KeyManagerStateSql], if none are found, it will return an empty `Vec`.
```

File: base_layer/wallet/src/key_manager_service/storage/sqlite_db/key_manager_state.rs [hunk 13]
```diff
@@ -85,17 +88,22 @@ impl NewKeyManagerStateSql {
...
+    /// Retrieve the key manager for the provided branch
+    /// Will return Err if the branch does not exist in the database
     pub fn get_state(branch: &str, conn: &SqliteConnection) -> Result<KeyManagerStateSql, KeyManagerStorageError> {
         key_manager_states::table
             .filter(key_manager_states::branch_seed.eq(branch.to_string()))
             .first::<KeyManagerStateSql>(conn)
             .map_err(|_| KeyManagerStorageError::KeyManagerNotInitialized)
     }
 
+    /// Update the key manager in the database to the one in memory
```
Comment (sdbondi) [inline base_layer/wallet/src/key_manager_service/storage/sqlite_db/key_manager_state.rs]:
```suggestion
    /// Creates or updates the database with the key manager state in this instance.
```

**Review by stringhandler — DISMISSED** [review]

**Comment by sdbondi** [comment]
@SWvheerden Need to run cargo fmt

**Review by stringhandler — APPROVED** [review]
